DOC-734 Adds redpanda_common and redpanda inputs and outputs and redpanda config updates (#97)
redpanda-data/rp-connect-docs: modules/components/pages/inputs/redpanda_common.adoc

      username: foo

```

== Consumer groups

When you specify a consumer group in your configuration, this input consumes one or more topics and automatically balances the topic partitions across any other connected clients with the same consumer group. Otherwise, topics are consumed in their entirety or with explicit partitions.

=== Delivery guarantees

If you choose to use consumer groups, the offsets of records received by Redpanda Connect are committed automatically. In the event of restarts, this input uses the committed offsets to resume data consumption where it left off.

Redpanda Connect guarantees at-least-once delivery. Records are only confirmed as delivered when all downstream outputs that a record is routed to have also confirmed delivery.

== Ordering

To preserve the order of topic partitions:

* Records consumed from each partition are processed and delivered in the order that they are received
* Only one batch of records of a given partition is processed at a time

This approach means that although records from different partitions may be processed in parallel, records from the same partition are processed in sequential order.

=== Delivery errors

> The order in which records are delivered may be disrupted by delivery errors and any error-handling mechanisms that start up. Redpanda Connect leans towards at-least-once delivery unless instructed otherwise, and this includes reattempting delivery of data when the ordering of that data is no longer guaranteed.

Reviewer: ```suggestion
The order in which records are delivered may be disrupted by delivery errors and any error-handling mechanisms that start up. Redpanda Connect uses at-least-once delivery unless instructed otherwise, and this includes reattempting delivery of data when the ordering of that data is no longer guaranteed.
```

Author: Updated.

*Type*: `array`

```yml
# Examples

topics:
  - foo
  - bar

topics:
  - things.*

topics:
  - foo,bar

topics:
  - foo:0
  - bar:1
  - bar:3

topics:
  - foo:0,bar:1,bar:3

topics:
  - foo:0-5
```

=== `regexp_topics`

Whether listed topics are interpreted as regular expression patterns for matching multiple topics. When topics are specified with explicit partitions, this field must remain set to `false`.

*Type*: `bool`

*Default*: `false`

=== `rack_id`

A rack specifies where the client is physically located, and changes fetch requests to consume from the closest replica as opposed to the leader replica.

*Type*: `string`

*Default*: `""`

=== `start_from_oldest`

Whether to consume from the oldest available offset. Otherwise, messages are consumed from the latest offset. This setting is applied when creating a new consumer group or the saved offset no longer exists.

*Type*: `bool`

*Default*: `true`

=== `fetch_max_bytes`

> The maximum amount of bytes that a broker tries to send during a fetch. 

Reviewer: ```suggestion
The maximum number of bytes that a broker tries to send during a fetch. 
```

Author: Updated.

If individual records are larger than the `fetch_max_bytes` value, brokers will still send them.

*Type*: `string`

*Default*: `50MiB`

=== `fetch_min_bytes`

> The minimum amount of bytes that a broker tries to send during a fetch. This field is equivalent to the Java setting `fetch.min.bytes`.

Reviewer: ```suggestion
The minimum number of bytes that a broker tries to send during a fetch. This field is equivalent to the Java setting `fetch.min.bytes`.
```

Author: Updated.

*Type*: `string`

*Default*: `1B`

=== `fetch_max_partition_bytes`

> The maximum amount of bytes that are consumed from a single partition in a fetch request. This field is equivalent to the Java setting `fetch.max.partition.bytes`.

Reviewer: ```suggestion
The maximum number of bytes that are consumed from a single partition in a fetch request. This field is equivalent to the Java setting `fetch.max.partition.bytes`.
```

Reviewer: I believe amount is for items that are uncountable, like raindrops in the sky. Number is for something countable...at least that's how I've always used them.

If a single batch is larger than the `fetch_max_partition_bytes` value, the batch is still sent so that the client can make progress. 

*Type*: `string`

*Default*: `1MiB`

=== `consumer_group`

An optional consumer group. When this value is specified: 

- The partitions of any topics, specified in the `topics` field, are automatically distributed across consumers sharing a consumer group
- Partition offsets are automatically committed and resumed under this name

Consumer groups are not supported when you specify explicit partitions to consume from in the `topics` field.

*Type*: `string`

=== `commit_period`

The period of time between each commit of the current partition offsets. Offsets are always committed during shutdown.

*Type*: `string`

*Default*: `5s`

=== `partition_buffer_bytes`

A buffer size (in bytes) for each consumed partition, which allows the internal queuing of records before they are flushed. Increasing this value may improve throughput but results in higher memory utilization. 

Each buffer can grow slightly beyond this value.

*Type*: `string`

*Default*: `1MB`

> === `auto_replay_nacks`

Reviewer: "nacked" is not a word as far as I can tell--it's not in the 3 dictionaries I checked. We should refrain from introducing "non-words" or jargon that is not intuitive to our users. Can this be changed?

= redpanda_common
// tag::single-source[]
:type: input
:page-beta: true
:categories: ["Services"]

component_type_dropdown::[]

Consumes data from a Redpanda (Kafka) broker, using credentials from a common `redpanda` configuration block.

ifndef::env-cloud[]
Introduced in version 4.39.0.
endif::[]

> Use this connector in conjunction with the xref:components:outputs/redpanda_common.adoc[`redpanda_common` output], to read data from one or more topics on a Redpanda cluster, and then write the data to another topic on the same cluster.

Author: Is this correct? Just trying to make it easier for users to understand when/if to use this component.

Reviewer 2: Addressed above

Author: Removed.

[tabs]
======
Common::
+
--

```yml
# Common configuration fields, showing default values
input:
  label: ""
  redpanda_common:
    topics: [] # No default (required)
    regexp_topics: false
    consumer_group: "" # No default (optional)
    auto_replay_nacks: true
```

--
Advanced::
+
--

```yml
# All configuration fields, showing default values
input:
  label: ""
  redpanda_common:
    topics: [] # No default (required)
    regexp_topics: false
    rack_id: ""
    start_from_oldest: true
    fetch_max_bytes: 50MiB
    fetch_min_bytes: 1B
    fetch_max_partition_bytes: 1MiB
    consumer_group: "" # No default (optional)
    commit_period: 5s
    partition_buffer_bytes: 1MB
    auto_replay_nacks: true
```

--
======

== Pipeline example

This data pipeline reads data from `topic_A` on a Redpanda cluster, transforms the data, and then writes it to `topic_B` on the same cluster. The cluster details are configured within the `redpanda` configuration block, so you only need to configure them once. This is a useful feature when you have multiple inputs and outputs in the same data pipeline that need to connect to the same cluster.

> ```

Author: If my understanding of the use case for this connector is correct, it would be good to add an example using a `redpanda_common` input and output + the config block. Could you provide this?

Reviewer 2: ```
input:
  redpanda_common:
    topics: [ foo, bar ]

output:
  redpanda_common:
    topic: baz
    key: ${! @kafka_key }

redpanda:
  seed_brokers: [ "127.0.0.1:9092" ]
  tls:
    enabled: true
  sasl:
    - mechanism: SCRAM-SHA-512
      password: bar
      username: foo
```

This approach means that although records from different partitions may be processed in parallel, records from the same partition are processed in sequential order.

=== Delivery errors

The order in which records are delivered may be disrupted by delivery errors and any error-handling mechanisms that start up. Redpanda Connect leans towards at-least-once delivery unless instructed otherwise, and this includes reattempting delivery of data when the ordering of that data is no longer guaranteed.

For example, a batch of records is sent to an output broker and only a subset of records are delivered. In this scenario, Redpanda Connect (by default) attempts to deliver the records that failed, even though these delivery failures may have been sent before records that were delivered successfully.

==== Use a fallback output

To prevent delivery errors from disrupting the order of records, you must specify a xref:components:outputs/fallback.adoc[`fallback`] output in your pipeline configuration. When adding a `fallback` output, it is good practice to set the `auto_retry_nacks` field to `false`. This also improves the throughput of your pipeline. 

> For example, the following configuration includes a `fallback` output that writes delivery errors into a dead letter queue topic (`foo_dlq`), which is also retried indefinitely as a way to apply back pressure.

Reviewer 2: If a message fails to be delivered to the first output within the `fallback` then the next output in the list is attempt, which in this case is another topic.

> For example, the following configuration includes a `fallback` output that writes delivery errors into a dead letter queue topic (`foo_dlq`), which is also retried indefinitely as a way to apply back pressure.

Author: Updated.
